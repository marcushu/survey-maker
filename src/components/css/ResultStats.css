.percentages {
    background-color: rgba(243, 66, 66, 0.472);
    color: white;
    padding: 3px;
    border-radius: 2px;
}
.percentagesGreen {
    background-color: rgb(44, 144, 82);
    color: white;
    padding: 3px;
    border-radius: 2px;
}

.question {
    font-weight: bold;
}

.dataRow {
    padding: 10px;
    border-bottom: 1px;
    border-bottom-style: solid;
    border-bottom-color:  rgb(1, 85, 134);
}