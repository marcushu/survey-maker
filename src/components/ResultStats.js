import React from 'react';
import { useRef } from 'react';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import './css/ResultStats.css'
import PropTypes from 'prop-types';


export default function ResultStats({ surveyData }) {
  const numberOfAnswers = useRef();

  const countAnswers = answerObj => {
    let answerCount = [];

    for (let k in answerObj) {
      // we're only interested in mult. choice results (who's 
      // keys are not known untill run-time)
      if ( k !== "question"
        && k !== "questionType"
        && k !== ""
        && answerObj.questionType !== "TEXT")
          answerCount.push({ answer: k, count: answerObj[k] })
    }

    // use this as the denomiator to calculate percentages
    numberOfAnswers.current = answerCount
      .map(({ count }) => count)
      .reduce((acc, curr) => acc + curr);

    return answerCount;
  }


  const percent = numerator => Math.round((numerator / numberOfAnswers.current) * 100);


  return (
    <>
      {surveyData.filter(res => res.questionType !== "TEXT").map((res, index) =>
        <Row className="dataRow" key={index}>
          <Col>
            {res.question}
          </Col>
          {countAnswers(res).map(countObj =>
            <>
              <Col md={1}>
                <span className="question">
                  {countObj.answer.replace('_', " ")}
                </span>
              </Col>
              <Col md={1}>
                <span
                  className={percent(countObj.count) >= 50 ? "percentagesGreen" : "percentages"}
                  data-testid="percentResult">
                  {percent(countObj.count)}%
                </span>
              </Col>
            </>)}
        </Row>)}
    </>
  )
}

ResultStats.propTypes = {
  surveyData: PropTypes.array
}