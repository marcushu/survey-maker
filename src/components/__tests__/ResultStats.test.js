import React from 'react';
import { render } from '@testing-library/react';
import ResultStats from '../ResultStats';

describe("resultStats", () => {
    const mockData = [
        {
            No: 1,
            Yes: 3,
            question: "did you attend",
            questionType: "BINARY"
        },
        {
            "12 mostly family": 1,
            "asd asdf sad fsd ": 1,
            "asd asdf sadf asdf ": 1,
            asdfasfd: 1,
            question: "how many in your party",
            questionType: "TEXT"
        },
        {
            good: 1,
            great: 3,
            question: "the event was...",
            questionType: "MULTCHOICE"
        },
        {
            No: 2,
            Yes: 2,
            question: "I had fun.",
            questionType: "BINARY"
        }
    ]

    it("renders", () => {
        const { asFragment } = render(
            <ResultStats surveyData={mockData} />
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("renders the questions", () => {
        const { getByText } = render(
            <ResultStats surveyData={mockData} />
        );

        expect(getByText(/did you attend/i)).toBeInTheDocument();
        expect(getByText(/the event was.../i)).toBeInTheDocument();
        expect(getByText(/I had fun./i)).toBeInTheDocument();
    });

    it("sets the correct display class for results above/below %50", () => {
        const { container } = render(
            <ResultStats surveyData={mockData} />
        );

        // find any result displayed as > %50
        const pctResultHigh = container.getElementsByClassName("percentagesGreen")[0].innerHTML;
        // find any result displayed as < %50
        const pctResultLow = container.getElementsByClassName("percentages")[0].innerHTML;

        expect(pctResultHigh.slice(0, -1) > 50).toBeTruthy();
        expect(pctResultLow.slice(0, -1) < 50).toBeTruthy();

    })
})