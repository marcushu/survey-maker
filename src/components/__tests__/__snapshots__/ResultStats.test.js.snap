// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resultStats renders 1`] = `
<DocumentFragment>
  <div
    class="dataRow row"
  >
    <div
      class="col"
    >
      did you attend
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="question"
      >
        No
      </span>
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="percentages"
        data-testid="percentResult"
      >
        25%
      </span>
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="question"
      >
        Yes
      </span>
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="percentagesGreen"
        data-testid="percentResult"
      >
        75%
      </span>
    </div>
  </div>
  <div
    class="dataRow row"
  >
    <div
      class="col"
    >
      the event was...
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="question"
      >
        good
      </span>
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="percentages"
        data-testid="percentResult"
      >
        25%
      </span>
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="question"
      >
        great
      </span>
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="percentagesGreen"
        data-testid="percentResult"
      >
        75%
      </span>
    </div>
  </div>
  <div
    class="dataRow row"
  >
    <div
      class="col"
    >
      I had fun.
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="question"
      >
        No
      </span>
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="percentagesGreen"
        data-testid="percentResult"
      >
        50%
      </span>
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="question"
      >
        Yes
      </span>
    </div>
    <div
      class="col-md-1"
    >
      <span
        class="percentagesGreen"
        data-testid="percentResult"
      >
        50%
      </span>
    </div>
  </div>
</DocumentFragment>
`;
